### PR TITLE
CXP-1413: Map product family (string)

### PR DESCRIPTION
**Pull request by eoliveauakeneo**
**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

- [ ] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: components/catalogs/back/tests/Integration/Infrastructure/Validation/ProductMapping/AttributeSource/SystemAttributeFamilySourceTest.php [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+
+declare(strict_types=1);
+
+namespace Akeneo\Catalogs\Test\Integration\Infrastructure\Validation\ProductMapping\AttributeSource;
+
+use Akeneo\Catalogs\Infrastructure\Validation\ProductMapping\AttributeSource\SystemAttributeFamilySource;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+/**
+ * @copyright 2022 Akeneo SAS (http://www.akeneo.com)
```
Comment (guillaumefer-akeneo) [inline components/catalogs/back/tests/Integration/Infrastructure/Validation/ProductMapping/AttributeSource/SystemAttributeFamilySourceTest.php:11]:
```suggestion
 * @copyright 2023 Akeneo SAS (http://www.akeneo.com)
```

**Review by guillaumefer-akeneo — APPROVED** [review]

File: components/catalogs/back/src/Infrastructure/Persistence/Family/GetFamilyLabelByCodeAndLocaleQuery.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+use Akeneo\Tool\Component\StorageUtils\Repository\SearchableRepositoryInterface;
+
+/**
+ * @copyright 2023 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+final class GetFamilyLabelByCodeAndLocaleQuery implements GetFamilyLabelByCodeAndLocaleQueryInterface
+{
+    /** @var array<string,array<string,string>> $familyLabelByCode */
+    private array $familyLabelByCode = [];
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/Persistence/Family/GetFamilyLabelByCodeAndLocaleQuery.php:18]: Normally, it's the same instance of the service that is injected in all places but to be completely sure your cache is efficient, make it static.

```suggestion
    private static array $familyLabelByCode = [];
```

File: components/catalogs/back/src/Infrastructure/Persistence/Family/GetFamilyLabelByCodeAndLocaleQuery.php [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+            $family->setLocale($locale);
+            $label = $family->getLabel() ?: $label;
+        }
+
+        if (!isset($this->familyLabelByCode[$code])) {
+            $this->familyLabelByCode[$code] = [];
+        }
+        $this->familyLabelByCode[$code][$locale] = $label;
+
+        return $label;
```
Comment (masacc) [inline components/catalogs/back/src/Infrastructure/Persistence/Family/GetFamilyLabelByCodeAndLocaleQuery.php:47]: the common way to do this is not to use an early return but rather like this:
```suggestion
        if (!isset($this->familyLabelByCode[$code]) || !isset($this->familyLabelByCode[$code][$locale])) {         
            $families = $this->searchableFamilyRepository->findBySearch(null, ['identifiers' => [$code]]);

            $label = \sprintf('[%s]', $code);

           if ([] !== $families) {
                /** @var FamilyInterface $family */
                $family = $families[0];
                $family->setLocale($locale);
                $label = $family->getLabel() ?: $label;
            }

            if (!isset($this->familyLabelByCode[$code])) {
                $this->familyLabelByCode[$code] = [];
            }
            $this->familyLabelByCode[$code][$locale] = $label;
        }
        
        return $this->familyLabelByCode[$code][$locale];
```

**Review by masacc — APPROVED** [review]

File: components/catalogs/back/src/Infrastructure/Persistence/Family/GetFamilyLabelByCodeAndLocaleQuery.php (hunk 2 above)
Comment (eoliveauakeneo) [reply components/catalogs/back/src/Infrastructure/Persistence/Family/GetFamilyLabelByCodeAndLocaleQuery.php:18]: Talked via slack, we will keep the non static array

**Pull request merged**